design/storage-class-device-set.md

```diff
@@ -0,0 +1,493 @@
+                        - cluster1
+                    topologyKey: "failure-domain.beta.kubernetes.io/zone"
+          volumeClaimTemplates:
+          - resources:
+              requests:
+                storage: 5Ti
+            storageClassName: gp2-ebs
+    - name: cluster1-set2
+      count: 3
+      resources:
+        requests:
+          cpu: 2
+          memory: 8Gi
+      placement:
+        podAntiAffinity:
+          preferredDuringSchedulingIgnoredDuringExecution:
+          - weight: 100
+            podAffinityTerm:
+              labelSelector:
+                matchExpressions:
+                - key: "rook.io/cluster"
+                  operator: In
+                  values:
+                    - cluster1
+                topologyKey: "failure-domain.beta.kubernetes.io/zone"
+      volumeClaimTemplates:
+      - resources:
+          requests:
+            storage: 10Ti
+        storageClassName: gp2-ebs
+```
+
+In this example, we have two `StorageClassDeviceSets` with different capacities
+for the devices in each set. The devices with larger capacities would require a
+greater amount of memory to operate the OSD Pods, so that is reflected in the
+`resources` field.
+
+#### Example 4: Simple local storage
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    nodes:
+    - name: node1
+    - name: node2
+    - name: node3
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      volumeClaimTemplates:
+      - resources:
+          requests:
+            storage: 1
+        storageClassName: cluster1-local-storage
+```
+
+In this example, we've defined that we already have three nodes. The nodes would
+be configured with one local storage device each, but they would not be
+specified in the `nodes` list. Prior to this, the admin would have had to deploy
+the local-storage-provisioner, created local PVs for each of the devices, and
+created a StorageClass to allow binding of the PVs to PVCs. At this point, the
+same workflow would be the same as the cloud use case, where you simply specify
+a count of devices, and a template with a StorageClass. Two notes here:
+
+1. The count of devices would need to match the number of existing devices to
+consume them all.
+1. The capacity for each device is irrelevant, since we will simply consume the
+entire storage device and get that capacity regardless of what is set for the
+PVC.
+
+#### Example 5: Multiple devices per OSD
+
+```yaml
+type: CephCluster
+name: cluster1
+...
+spec:
+  ...
+  storage:
+    ...
+    nodes:
+    - name: node1
+      devices:
+      - name: nvme01
+      - name: sdb
+    - name: node2
+      devices:
+      - name: nvme01
+      - name: sdb
+    - name: node3
+      devices:
+      - name: nvme01
+      - name: sdb
+    storageClassDeviceSet:
+    - name: cluster1-set1
+      count: 3
+      config:
+        metadataDevice: "/dev/rook/device1"
+      volumeClaimTemplates:
```

Comment:
Instead of having an array of templates, we just name them accordingly for the scenarios we care about? I'm thinking `dataTemplate` and `metadataTemplate` would suffice. If they didn't specify `metadataTemplate`, the osd would just be on a single device with a single PVC. And for this advanced scenario, we would need to create a single osd prepare job to provision them with a single ceph-volume command.

Reply:
This is also where the commonality can break down between storage providers. Others besides OSDs might not have use of `metadataTemplate`, in which case they could just ignore it.

Reply:
Even the notion of calling something a `dataTemplate` feels like it's getting too tied down to how Rook-Ceph intends to use it. The list of templates is much more generic.

Reply:
Now I see that we can set the `name` of each PVC template and storage providers could key off certain names.

Reply:
Having `name` key off sounds good so let's stick with @jarrpa current proposal.